controller/engine_image_controller.go

```diff
@@ -394,72 +394,67 @@ func (ic *EngineImageController) syncNodeDeploymentMap(engineImage *longhorn.Eng
 
 // handleAutoUpgradeEngineImageToDefaultEngineImage automatically upgrades volume's engine image to default engine image when it is applicable
 func (ic *EngineImageController) handleAutoUpgradeEngineImageToDefaultEngineImage() error {
-	concurrentAutomaticEngineUpgradePerNodeLimit, err := ic.ds.GetSettingAsInt(types.SettingNameConcurrentAutomaticEngineUpgradePerNodeLimit)
+	defaultEngineImage, err := ic.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
 	if err != nil {
 		return err
 	}
-	if concurrentAutomaticEngineUpgradePerNodeLimit <= 0 {
+	defaultEngineImageResource, err := ic.ds.GetEngineImage(types.GetEngineImageChecksumName(defaultEngineImage))
+	if err != nil {
+		return err
+	}
+	// Make sure that only the owner of the default engine image does the upgrade
+	if defaultEngineImageResource.Status.OwnerID != ic.controllerID {
 		return nil
 	}
 
-	defaultEngineImage, err := ic.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
+	concurrentAutomaticEngineUpgradePerNodeLimit, err := ic.ds.GetSettingAsInt(types.SettingNameConcurrentAutomaticEngineUpgradePerNodeLimit)
 	if err != nil {
 		return err
 	}
-	defaultEngineImageResource, err := ic.ds.GetEngineImage(types.GetEngineImageChecksumName(defaultEngineImage))
-	if err != nil {
-		return err
+	if concurrentAutomaticEngineUpgradePerNodeLimit <= 0 {
+		return nil
 	}
 
 	// List all volumes and select a set of volume for upgrading.
 	volumes, err := ic.ds.ListVolumes()
 	if err != nil {
 		return err
 	}
-	upgradingCandidates, err := ic.getVolumesForEngineImageUpgrading(volumes, defaultEngineImageResource)
-	if err != nil {
-		return err
-	}
 
-	// Make sure not to upgrade too many volumes on the same node at the same time.
-	upgradingCandidates = limitAutomaticEngineUpgradePerNode(volumes, upgradingCandidates, concurrentAutomaticEngineUpgradePerNodeLimit)
+	candidates, inProgress := ic.getVolumesForEngineImageUpgrading(volumes, defaultEngineImageResource)
 
-	// Upgrade engine image for volume
-	for _, v := range upgradingCandidates {
-		oldEngineImage := v.Spec.EngineImage
-		v.Spec.EngineImage = defaultEngineImage
-		v, err = ic.ds.UpdateVolume(v)
-		if err != nil {
-			return err
+	limitedCandidates := limitAutomaticEngineUpgradePerNode(candidates, inProgress, int(concurrentAutomaticEngineUpgradePerNodeLimit))
+
+	for _, vs := range limitedCandidates {
+		for _, v := range vs {
+			oldEngineImage := v.Spec.EngineImage
+			v.Spec.EngineImage = defaultEngineImage
+			v, err = ic.ds.UpdateVolume(v)
+			if err != nil {
+				return err
+			}
+			ic.logger.Infof("automatically upgrades volume's engine image from %v to the default engine image %v for volume %v", oldEngineImage, defaultEngineImage, v.Name)
```

Comment:
```suggestion
			log := ic.logger.withField("volume", v.Name).withField("engineImage", v.Spec.EngineImage).Infof("automatically upgrade volume engine image to the default engine image %v", defaultEngineImage)
			v.Spec.EngineImage = defaultEngineImage
			v, err = ic.ds.UpdateVolume(v)
			if err != nil {
				return err
			}
```

Reply:
using fields allows to trace on a specific volume/engineImage

Reply:
Thank you! Fixed